WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingViewModel.kt

```diff
@@ -1,13 +1,199 @@
+    data class TagsState(val tags: String? = null) : PrepublishingScreenState(TAGS)
+
+    // TODO add the other values for this state object.
+    @Parcelize
+    data class ActionsState(val tags: String? = null) : PrepublishingScreenState(HOME)
+}
+
+data class PrepublishingNavigationTarget(
+    val site: SiteModel,
+    val targetScreen: PrepublishingScreen,
+    val screenState: PrepublishingScreenState?
+)
+
 class PrepublishingViewModel @Inject constructor() : ViewModel() {
     private var isStarted = false
+    private lateinit var site: SiteModel
+    private var postRepository: EditPostRepository? = null
+    private var tagsState: TagsState? = null
+    private var currentScreenState: Pair<PrepublishingScreen, PrepublishingScreenState?>? = null
+
+    private val _navigationTarget = MutableLiveData<Event<PrepublishingNavigationTarget>>()
+    val navigationTarget: LiveData<Event<PrepublishingNavigationTarget>> = _navigationTarget
+
+    private val _dismissBottomSheet = MutableLiveData<Event<Unit>>()
+    val dismissBottomSheet: LiveData<Event<Unit>> = _dismissBottomSheet
 
-    fun start() {
+    fun start(
+        postRepository: EditPostRepository?,
+        site: SiteModel,
+        screenState: PrepublishingScreenState?
+    ) {
         if (isStarted) return
         isStarted = true
+
+        this.site = site
+        this.postRepository = postRepository
+
+        /**
+         * Restores the specific state if it was persisted.
+         */
+        screenState?.let {
+            restoreSavedScreenState(screenState)
+            updateScreenStatesFromPostRepository()
+            navigateToScreen(screenState.prepublishingScreen)
+        } ?: run {
+            updateScreenStatesFromPostRepository()
+            navigateToScreen(HOME)
+        }
+    }
+
+    /**
+     * This exists because any specific screen could be in the process of making changes to a value and process death
+     * occurs. So this is to restore the specific screen state that is impacted by this process.
+     */
+    private fun restoreSavedScreenState(screenState: PrepublishingScreenState) {
+        when (screenState) {
+            is TagsState -> tagsState = screenState
+        }
+        // TODO all the other screens that can be persisted.
+    }
+
+    private fun navigateToScreen(prepublishingScreen: PrepublishingScreen) {
+        when (prepublishingScreen) {
+            HOME -> updateNavigationTarget(PrepublishingNavigationTarget(site, HOME, ActionsState(tagsState?.tags)))
+            TAGS -> updateNavigationTarget(
+                    PrepublishingNavigationTarget(
+                            site,
+                            prepublishingScreen,
+                            tagsState
+                    )
+            )
+            PUBLISH -> TODO()
+            VISIBILITY -> TODO()
+        }
+    }
+
+    /**
+     * Checks to see if state was restored for a specific action that was ongoing. If not, then the Post Repository
```

Comment:
> The main point of the repository is to have a single source of truth - is there ever a point in time when the tagsState is out of sync with the repo?

I had created this because I was thinking that the save operation should only be triggered when the back button is pressed but honestly, I wasn't too fond of the approach; I eagerly awaited this review to verify my assumptions. Thanks. 

> Do we really need to have the tagsState field? Can't we always access the EditPostRepository? I assume we always want to persist the data ASAP right?

Yes, you are correct. It also makes fetching the data on other screens easier since the data is persisted and when the user navigates to the `Post Settings` the tags will be there. Thanks for this observation 🙏 

removed this here 15171b1e1702bbb537775965c3f05d577331573e

Reply:
❤️